tools/nl/embeddings/finetuning/finetune.py

```diff
@@ -0,0 +1,276 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Fine Tune the base model."""
+
+from datetime import datetime
+import glob
+import os
+import time
+from typing import Any, List
+
+from absl import app
+from absl import flags
+from google.cloud import storage
+import gspread
+import pandas as pd
+from sentence_transformers import InputExample
+from sentence_transformers import losses
+from sentence_transformers import SentenceTransformer
+from torch.utils.data import DataLoader
+import utils
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string('model_name_v2', 'all-MiniLM-L6-v2', 'Model name')
+flags.DEFINE_string('bucket_name_v2', 'datcom-nl-models', 'Storage bucket')
+flags.DEFINE_string('embeddings_size', '',
+                    'Valid values are: "small", "medium"')
+
+flags.DEFINE_string(
+    'autogen_input_basedir', 'data/autogen_input',
+    'Base path for CSVs with autogenerated SVs with name and description. '
+    'The actual path is `{--autogen_input_base}/{--embeddings_size}/*.csv`.')
+
+flags.DEFINE_string('alternatives_filepattern', 'data/alternatives/*.csv',
+                    'File pattern (relative) for CSVs with alternatives')
+
+flags.DEFINE_string('sentence_pairs_filepath',
+                    'data/finetuning/sentence_pairs.csv',
+                    'File path for sentence pairs CSV')
+
+BATCH_SIZE = 256
+NUM_WARMUP_STEPS = 10
+NUM_EPOCHS = 10
```

Comment:
Am I not seeing some commits on GH? It seems you intended to add some comments and refactor suggestion in fine_tune_model?

Reply:
So I did the refactoring but perhaps you were hoping it would go farther? Specifically, I added the function `_save_finetuned_model()` which performs step 3. Step 1 is just getting the alternatives and loading the base model. And Step 2 is the finetuning step, invoking the function `fine_tune_model()`.

Reply:
Oh sorry, I meant the _generate_training_examples() comment.  But that's just a nice-to-have, need not block merge.